tests/test_index.py

```diff
@@ -1153,3 +1153,96 @@ def test_compact_json(index_data):
+
+    # The function under test is looking for a package with features, and an
+    # older version of the same package without features.
+    with conn:  # transaction
+        conn.execute(
+            f"""INSERT INTO index_json VALUES('{features_pkg}','{{"build":"h39de5ba_0","build_number":0,"depends":[],"name":"{features_pkg_name}","noarch":"generic","subdir":"noarch","timestamp":1561127261940,"version":"1.0","md5":"ba68433ef44982170d4e2f2f9bf89338","sha256":"33877cbe447e8c7a026fbcb7e299b37208ad4bc70cf8328fb4cf552af01ada76","size":2683,"track_features":["jim"],"features":["jim"]}}');"""
+        )
+        conn.execute(
+            f"""INSERT INTO stat VALUES('indexed','{features_pkg}',1652905054,2683,'33877cbe447e8c7a026fbcb7e299b37208ad4bc70cf8328fb4cf552af01ada76','ba68433ef44982170d4e2f2f9bf89338',NULL,NULL);"""
+        )
+        conn.execute(
+            f"""INSERT INTO stat VALUES('fs','{features_pkg}',1652905054,2683,NULL,NULL,NULL,NULL);"""
+        )
+
+        conn.execute(
+            f"""INSERT INTO index_json VALUES('{features_pkg_2}','{{"build":"h39de5ba_0","build_number":0,"depends":[],"name":"{features_pkg_name}","noarch":"generic","subdir":"noarch","timestamp":1561127261940,"version":"0.9","md5":"ba68433ef44982170d4e2f2f9bf89338","sha256":"33877cbe447e8c7a026fbcb7e299b37208ad4bc70cf8328fb4cf552af01ada76","size":2683}}');"""
+        )
+        conn.execute(
+            f"""INSERT INTO stat VALUES('indexed','{features_pkg_2}',1652905054,2683,'33877cbe447e8c7a026fbcb7e299b37208ad4bc70cf8328fb4cf552af01ada76','ba68433ef44982170d4e2f2f9bf89338',NULL,NULL);"""
+        )
+        conn.execute(
+            f"""INSERT INTO stat VALUES('fs','{features_pkg_2}',1652905054,2683,NULL,NULL,NULL,NULL);"""
+        )
+
+        # cover "run exports on .conda package" branch
+        conn.execute(f"""INSERT INTO run_exports VALUES('{features_pkg_2}','{{}}')""")
+
+        # cover "check for unknown file extension" branch
+        conn.execute(
+            """INSERT INTO stat VALUES('fs','unexpected-filename',1652905054,2683,NULL,NULL,NULL,NULL)"""
+        )
+        conn.execute("""INSERT INTO index_json VALUES('unexpected-filename','{}');""")
+
+    # Call internal "write repodata.json" function normally called by
+    # channel_index.index(). index_prepared_subdir doesn't check which packages
+    # exist.
+    channel_index.index_prepared_subdir("noarch", False, False, None, None)
+
+    # complain about 'unexpected-filename'
+    channel_index.build_run_exports_data("noarch")
+
+
+def test_bad_patch_version(index_data):
+    """
+    Test unsupported patches.
+    """
+    pkg_dir = Path(index_data, "packages")
+
+    # compact json
+    channel_index = conda_index.index.ChannelIndex(
+        str(pkg_dir),
+        None,
+        write_bz2=False,
+        write_zst=False,
+        compact_json=True,
+        threads=1,
+    )
+
+    instructions = Path(__file__).parents[1] / "tests" / "gen_patch_2.py"
```

Comment:
```suggestion
    instructions = Path(__file__).parent / "gen_patch_2.py"
```

Reply:
Is `parents[0]` the same as `parent`

Reply:
why do we need the second parent? `gen_patch_2.py` is right next to this file in the same directory?

Reply:
I see what you mean.